### PR TITLE
Inline typeclass projections.

### LINKED ISSUE #5748: Inline known type class dictionaries
Avoid dictionary passing whenever the dictionary is known at compile time by inlining the type class method and the dictionary.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
TL;DR: This PR offers 1.10x to 1.15x speedups on our benchmarks, by removing layers of indirection when it comes to calling typeclass functions at known typeclasses. To do this, it also implements additional functions for manipulating the DAML-LF AST in Haskell (in particular substitution and alpha equivalence for expressions), which will be very useful for implementing future optimizations.

This PR:

* refactors free variable set calculation to a FreeVars module
* adds an Alpha module for calculation of alpha equivalence of LF expressions and types. Previously we only had alpha equivalenc for types, and it was incomplete (no case for `TSynApp`).
* adds a Subst module for substitution within LF expressions.
  This implementation piggybacks on the existing type
  substition.
* adds lots of tests for Alpha and Subst.
* passes World data into the simplifier, in order to have
  have the ability to inline functions.
    * Except when doing incremental builds, in which case the simplifier gets an empty World. This 100% prevents cross-module inlining when incremental builds are turned on. As @cocreature pointed out, incremental builds only rebuild whenever a module ABI changes, which means the implementation of module A could change without module B being recomputed, even if B depends on A, so it's unsafe to do any inlining in this case (or rely on implementations in any way).
* inlines and beta-reduces typeclass projection functions applied to typeclass dictionary functions, and so fixes #5748 (see sample results)
* provides 10-13% time reduction, i.e. 1.10x to 1.15x speedup, in the benchmarks.

Sample results:

DAML input:

```
module Main where

hello : Int
hello = 10 + 30
```

Original DAML-LF output:

```
module Main where
    @location(7:0-7:5)
    def hello : Int64 =
      a284919a95c4a515cd1efac0d89be302d0e9d61e692a2176128c871ad8067e36:GHC.Num:+
        @Int64
        a284919a95c4a515cd1efac0d89be302d0e9d61e692a2176128c871ad8067e36:GHC.Num:$fAdditiveInt
        10
        30
```

Current DAML-LF output:

```
 module Main where
    @location(7:0-7:5)
    def hello : Int64 = ADD_INT64 10 30
```

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    (typeVars0, exprVars0) = freeVarsInExpr e
+
+
+freeVarsInType :: Type -> Set.Set TypeVarName
+freeVarsInType = Type.freeVars
+
+freeVarsInTypeConApp :: TypeConApp -> Set.Set TypeVarName
+freeVarsInTypeConApp (TypeConApp _ tys) = foldMap freeVarsInType tys
+
+freeVarsInExpr :: Expr -> (Set.Set TypeVarName, Set.Set ExprVarName)
```
[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs:83]: They do sufficiently different things (`freeVarsStep` calculates only free `ExprVarNames`, and does it at the same time as calculating safeties).

[the remark this answers is not in the document]
Comment (hurryabit) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs:83]: Could we only have something like `freeVarsStep` for type variables here and combine it with `freeVarsStep` like in
https://github.com/digital-asset/daml/blob/f71ab92ea52bdc71aba6446264cbe8af61a7805f/compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs#L244

[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs:83]: Happy to do some coalescing / refactoring after I figure out a good way to reduce the janky "run simplifier twice" & fix other issues.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 2]
```diff
@@ -241,14 +200,118 @@ safetyStep = \case
...
+                | otherwise -> TCNeither
+
+    ETyAppF tci ty ->
+        case tci of
+            TCProjection (ETyLam (x,_) e) ->
+                TCProjection (substExpr (typeSubst x ty) e)
+            TCDictionary (ETyLam (x,_) e) ->
+                TCDictionary (substExpr (typeSubst x ty) e)
+            _ ->
+                TCNeither
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:235]: Shouldn't we allow `ETmAppF` for dictionaries as well in order to inline dictionaries like `instance Eq a => Eq [a]`?

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 3]
```diff
@@ -241,14 +200,118 @@ safetyStep = \case
...
+    e@(EStructCon _) ->
+        Just e
+
+    EVal x
+        | Right dv <- lookupValue x world
+        , isTypeClassDictionary dv
+        -> do
+            Just (dvalBody dv)
+
+    ETyApp e t -> do
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:285]: Hasn't `typeclassStep` already remove all type applications?

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 4]
```diff
@@ -241,14 +200,118 @@ safetyStep = \case
...
+        ETyLam (x,_) e' <- getTypeClassDictionary world e
+        Just (substExpr (typeSubst x t) e')
+
+    ETmApp e1 e2 -> do
+        ETmLam (x,_) e1' <- getTypeClassDictionary world e1
+        Just (substExpr (exprSubst x e2) e1')
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:291]: Aren't we running danger to duplicate code here? Should we avoid this with a `let` binding when necessary?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/FreeVars.hs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    ENoneF t -> freeVarsInType t
+    ESomeF t e -> freeVarsInType t <> e
+    EToAnyF t e -> freeVarsInType t <> e
+    EFromAnyF t e -> freeVarsInType t <> e
+    ETypeRepF t -> freeVarsInType t
+
+  where
+
+    goCase :: (CasePattern, FreeVars) -> FreeVars
+    goCase = snd -- overapproximation is fine and cheap
```
[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/FreeVars.hs:104]: 👍 Making it more precise.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    , substExhausted = freeExprVar x <> fvs
+    }
+
+typeSubst :: TypeVarName -> Type -> Subst
+typeSubst x t = typeSubst' x t (freeVarsInType t)
+
+exprSubst :: ExprVarName -> Expr -> Subst
+exprSubst x e = exprSubst' x e (freeVarsInExpr e)
+
+substType :: Subst -> Type -> Type
```
[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Subst.hs:70]: good idea 👍

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 4 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:291]: I don't think so. This is here because of typeclass specialization.

Introducing a let would make the optimization a lot more complicated -- can't apply typeclass projection on a let directly.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 3 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:285]: Nope, because of type class specialization.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 2 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:235]: Do these count as "known typeclasses"? AFAIK if you have a fixed `a`, GHC will specialize the typeclass, so we end up inlining them. But if you have a generic `a` then you aren't going to get inlining.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 7]
```diff
@@ -241,14 +200,118 @@ safetyStep = \case
...
+isTypeClassDictionary DefValue{..}
+    = T.isPrefixOf "$f" (unExprValName (fst dvalBinder)) -- generic dictionary
+    || T.isPrefixOf "$d" (unExprValName (fst dvalBinder)) -- specialized dictionary
 
-simplifyExpr :: Expr -> Expr
-simplifyExpr = fst . cata go
+isTypeClassProjection :: DefValue -> Bool
+isTypeClassProjection DefValue{..} = go dvalBody
   where
+    go :: Expr -> Bool
```
[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:211]: Yeah, I don't think we get anything by making it more precise right now. 🤷 The only structural projections are the ones we insert for typeclass desugaring.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 2 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:235]: I think we can perhaps leave this (& further experiments with inlining) to a separate PR.
